packages/babel-preset-default/index.js

```diff
@@ -1,36 +1,78 @@
 module.exports = function( api ) {
+	let wpBuildOpts = {};
 	const isTestEnv = api.env() === 'test';
 
+	// Because we serve a different preset depending
+	// on the caller options, we need to tell the cache
+	// when it needs updating. Otherwise, it won't be
+	// recalculated for different builds.
+	api.cache.using( () => wpBuildOpts.name );
```

Comment:
The babel cache needs to be updated every time the `wpBuildOpts.name` changes. We need to do this to make sure the preset is updated. The [`bin/packages/build.js` script](https://github.com/WordPress/gutenberg/blob/664781a47820d6c37c0ef8366ac9b9d293e28b13/bin/packages/build.js) currently iterates over each source file and generates both bundles (`main` and `modules`), so this means the cache won't be used.

I like to time `bin/packages/build.js` script to measure if this makes a noticeable difference.

Reply:
> I like to time `bin/packages/build.js` script to measure if this makes a noticeable difference.

Compared this branch to `master` and both take about the same time to execute `npm run dev` and `npm run build`.

Reply:
I can't see `api.cache` used in any of the Babel presets: https://github.com/babel/babel/search?p=3&q=api.cache&unscoped_q=api.cache

We probably should skip it as well.

Reply:
We are already using `api.env` in `babel-preset-default`, which updates the `api.cache` automatically for us. [See](https://babeljs.io/docs/en/config-files#apienv):

> This function internally makes use of api.cache mentioned below to ensure that Babel is aware that this build depends on a specific envName.

Once we introduce variable preset (different config options depending on what we want to build) we need to update the cache manually so it returns the proper results. We could perhaps rewrite the current approach to use node env variables and `api.env` but will be updating the cache in the preset anyway.

My reading of the situation is that we're fine here.

Reply:
So, it looks like the [`api.caller`](https://babeljs.io/docs/en/config-files#apicallercb) is also used by the cache to determine whether to re-compute the preset if it isn't already parsed. We don't need to manually update it, just returning the proper key.

For `npm run build` and `npm run dev`, I've checked that the preset is only parsed three times: for the keys `WP_BUILD_MAIN`, `WP_BUILD_MODULE`, and `undefined`. The first two are parsed when the `bin/packages/build.js` is executed. The other one happens when webpack calls `babel-loader` (which doesn't pass any `caller` option).

Note that `api.test` will also update the cache if the environment is different. This only happens when we run tests. I think we can perhaps simplify this to only use the `api.caller`, but I'd like to do that in a different PR.